File: basic/13-knapsack/knapsack.js
function knapsack(W, wt, val, n) {
  const K = Array.from({ length: n + 1 }, () => Array(W + 1).fill(0));

  for (let i = 0; i <= n; i++) {
      for (let w = 0; w <= W; w++) {
          if (i === 0 || w === 0) {
              K[i][w] = 0;
          } else if (wt[i - 1] <= w) {
              K[i][w] = Math.max(val[i - 1] + K[i - 1][w - wt[i - 1]], K[i - 1][w]);
          } else {
              K[i][w] = K[i - 1][w];
          }
      }
  }
  return K[n][W];
}

// 示例
const val = [60, 100, 120];
const wt = [10, 20, 30];
const W = 50;
const n = val.length;
const result = knapsack(W, wt, val, n);
console.log("最大价值为:", result);

File: basic/dynamic-programming/knapsack/knapsack.rb
def knapsack(W, wt, val, n)
  K = Array.new(n + 1) { Array.new(W + 1, 0) }

  (0..n).each do |i|
    (0..W).each do |w|
      if i == 0 || w == 0
        K[i][w] = 0
      elsif wt[i - 1] <= w
        K[i][w] = [val[i - 1] + K[i - 1][w - wt[i - 1]], K[i - 1][w]].max
      else
        K[i][w] = K[i - 1][w]
      end
    end
  end

  K[n][W]
end

# 示例
val = [60, 100, 120]
wt = [10, 20, 30]
W = 50
n = val.length
result = knapsack(W, wt, val, n)
puts "最大价值为: #{result}"
